Add EH info support to R2RDump

trylek opened the pull request:
This change expands runtime function dump to include the exception
handling info looked up via the EXCEPTION_INFO R2R header table.

Thanks

Tomas

zamont commented on src/tools/r2rdump/EHInfo.cs:
```diff
@@ -0,0 +1,234 @@
+        /// <summary>
+        /// Length of the serialized EH clause in the PE image.
+        /// </summary>
+        public const int Length = 6 * sizeof(int);
```
In the constructor you use `sizeof(uint)` but here you're using `sizeof(int)`. Do these two need to be consistent?

trylek replied:
Well, the sizes are obviously equal but I'll fix this for the sake of consistency, thanks for pointing that out!

zamont commented on src/tools/r2rdump/EHInfo.cs:
```diff
@@ -0,0 +1,234 @@
+            while ((length -= 2 * sizeof(uint)) >= 8)
+            {
+                offset += 2 * sizeof(uint);
+                int rva2 = BitConverter.ToInt32(image, offset);
```
I don't see rva2 being used anywhere. Is that intentional?

trylek replied:
Of course not, thanks for spotting! This is supposed to be a state machine iterating over successive pairs of records so that, at the end of the loop statement, we need to assign rva2 -> rva1, eh2 -> eh1. Will fix!

BruceForstall commented on src/tools/r2rdump/EHInfo.cs:
```diff
@@ -0,0 +1,237 @@
+        COR_ILEXCEPTION_CLAUSE_FILTER = 0x0001,         // If this bit is on, then this EH entry is for a filter
+        COR_ILEXCEPTION_CLAUSE_FINALLY = 0x0002,        // This clause is a finally clause
+        COR_ILEXCEPTION_CLAUSE_FAULT = 0x0004,          // Fault clause (finally that is called on exception only)
+        COR_ILEXCEPTION_CLAUSE_DUPLICATED = 0x0008,     // duplicated clause. This clause was duplicated to a funclet which was pulled out of line
```
corinfo.h also has:
```
CORINFO_EH_CLAUSE_SAMETRY   = 0x0010, // This clause covers same try block as the previous one. (Used by CoreRT ABI.)
```

trylek replied:
Not sure about relevance of this one as we're explicitly targeting the CoreCLR ABI with the CPAOT work.

BruceForstall commented on src/tools/r2rdump/EHInfo.cs:
```diff
@@ -0,0 +1,237 @@
+            {
+                stringBuilder.Append(" FAULT");
+            }
+            if ((Flags & CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_DUPLICATED) != CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_NONE)
```
I think these checks are slightly wrong. The definition is quite weird, since only one of NONE, FILTER, FINALLY, and FAULT can be true at a time. Thus, you can do a bit check. Or, you can simply mask 0x7 and direct compare. However, DUPLICATED can be set at the same time as one of the others (e.g., DUPLICATED as well as FILTER). So just do a bit check.

Seems like for NONE you should explicitly say it's a class handler, and maybe (if possible) construct the class name corresponding to the ClassToken.

trylek replied:
Thanks @BruceForstall, that's a very good suggestion, will fix!

trylek replied:
Fixed in 3rd commit.

trylek commented:
@BruceForstall PTAL - I believe I have addressed your feedback in the 3rd commit:

<pre>
EH info @ BD51, #clauses = 1
Flags 00 TryOff 001D (RVA 6B51) TryEnd 0047 (RVA 6B7B) HndOff 0051 (RVA 6B85) HndEnd 00C9 (RVA 6BFD) ClsFlt 1000011 CATCH: System.Exception
</pre>

BruceForstall commented on src/tools/r2rdump/EHInfo.cs:
```diff
@@ -98,21 +113,29 @@ public void WriteTo(StringBuilder stringBuilder, int methodRva)
+                    break;
+
+                case CorExceptionFlag.COR_ILEXCEPTION_CLAUSE_FINALLY:
+                    stringBuilder.AppendFormat(" FINALLY (RVA {0:X4})", ClassTokenOrFilterOffset + methodRva);
```
In the FINALLY and FAULT cases, the "ClassToken"/"FilterOffset" field is unused, so I don't think you should try to interpret it.

trylek commented:
@BruceForstall Ad DUPLICATED. It would be great if you had a chance to scrutinize the log.

<code>
    private static bool StreamReaderReadLine()
    {
        Console.WriteLine($@"Dumping file: {TextFileName}");
        using (StreamReader reader = new StreamReader(TextFileName, System.Text.Encoding.UTF8))
        {
            Console.WriteLine("StreamReader created ...");
            string line1 = reader.ReadLine();
            Console.WriteLine($@"Line 1: {line1}");
            string line2 = reader.ReadLine();
            Console.WriteLine($@"Line 2: {line2}");
            return line2 != null;
        }
    }
</code>

<pre>
Boolean Program.StreamReaderReadLine()
Token: 0x0600000B
Rid: 11
EntryPointRuntimeFunctionId: 24
Number of RuntimeFunctions: 2
Number of fixups: 8
    TableIndex 1, Offset 001E
    TableIndex 1, Offset 001F
    TableIndex 1, Offset 0041
    TableIndex 1, Offset 0042
    TableIndex 5, Offset 002E
    TableIndex 5, Offset 002F
    TableIndex 5, Offset 0030
    TableIndex 5, Offset 0031


Program.StreamReaderReadLine()
Id: 24
StartAddress: 0x00007828
Size: 336 bytes
UnwindRVA: 0x0000A5F0
Version:            1
Flags:              0x03 EHANDLER UHANDLER
SizeOfProlog:       0x0009
CountOfUnwindCodes: 4
FrameRegister:      RAX
FrameOffset:        0x0
PersonalityRVA:     0x6000
UnwindCode[0]: CodeOffset 0x0009 FrameOffset 0x0109 NextOffset 0x152 Op 0 - 
UnwindCode[1]: CodeOffset 0x0013 FrameOffset 0x0013 NextOffset 0x-1 Op Scaled small
UnwindCode[2]: CodeOffset 0x0002 FrameOffset 0x7002 NextOffset 0x-1 Op RDI(7)
UnwindCode[3]: CodeOffset 0x0001 FrameOffset 0x5001 NextOffset 0x-1 Op RBP(5)

EH info @ BDF9, #clauses = 2
Flags 02 TryOff 0096 (RVA 78BE) TryEnd 013D (RVA 7965) HndOff 0150 (RVA 7978) HndEnd 0189 (RVA 79B1) ClsFlt 0000 FINALLY
Flags 0A TryOff 013D (RVA 7965) TryEnd 013D (RVA 7965) HndOff 013D (RVA 7965) HndEnd 0146 (RVA 796E) ClsFlt 0000 FINALLY DUPLICATED

Debug Info
	Bounds:
	Native Offset: 0x0, IL Offset: 0xFFFFFFFE, Source Types: StackEmpty
	Native Offset: 0x22, IL Offset: 0x0, Source Types: StackEmpty
	Native Offset: 0x23, IL Offset: 0x1, Source Types: StackEmpty
	Native Offset: 0x46, IL Offset: 0xB, Source Types: CallInstruction
	Native Offset: 0x50, IL Offset: 0x10, Source Types: SourceTypeInvalid
	Native Offset: 0x54, IL Offset: 0x10, Source Types: CallInstruction
	Native Offset: 0x5A, IL Offset: 0x15, Source Types: StackEmpty
	Native Offset: 0x5B, IL Offset: 0x16, Source Types: StackEmpty
	Native Offset: 0x68, IL Offset: 0x1B, Source Types: CallInstruction
	Native Offset: 0x72, IL Offset: 0x20, Source Types: SourceTypeInvalid
	Native Offset: 0x88, IL Offset: 0x20, Source Types: CallInstruction
	Native Offset: 0x8E, IL Offset: 0x25, Source Types: SourceTypeInvalid
	Native Offset: 0x96, IL Offset: 0x26, Source Types: StackEmpty
	Native Offset: 0x97, IL Offset: 0x27, Source Types: StackEmpty
	Native Offset: 0xA1, IL Offset: 0x2C, Source Types: CallInstruction
	Native Offset: 0xA7, IL Offset: 0x31, Source Types: StackEmpty
	Native Offset: 0xA8, IL Offset: 0x32, Source Types: StackEmpty
	Native Offset: 0xB5, IL Offset: 0x33, Source Types: CallInstruction
	Native Offset: 0xBF, IL Offset: 0x38, Source Types: SourceTypeInvalid
	Native Offset: 0xC7, IL Offset: 0x39, Source Types: StackEmpty
	Native Offset: 0xD5, IL Offset: 0x3F, Source Types: CallInstruction
	Native Offset: 0xDF, IL Offset: 0x44, Source Types: SourceTypeInvalid
	Native Offset: 0xE3, IL Offset: 0x44, Source Types: CallInstruction
	Native Offset: 0xE9, IL Offset: 0x49, Source Types: StackEmpty
	Native Offset: 0xEA, IL Offset: 0x4A, Source Types: StackEmpty
	Native Offset: 0xF7, IL Offset: 0x4B, Source Types: CallInstruction
	Native Offset: 0x101, IL Offset: 0x50, Source Types: SourceTypeInvalid
	Native Offset: 0x109, IL Offset: 0x51, Source Types: StackEmpty
	Native Offset: 0x117, IL Offset: 0x57, Source Types: CallInstruction
	Native Offset: 0x121, IL Offset: 0x5C, Source Types: SourceTypeInvalid
	Native Offset: 0x125, IL Offset: 0x5C, Source Types: CallInstruction
	Native Offset: 0x12B, IL Offset: 0x61, Source Types: StackEmpty
	Native Offset: 0x12C, IL Offset: 0x62, Source Types: StackEmpty
	Native Offset: 0x13A, IL Offset: 0x67, Source Types: StackEmpty
	Native Offset: 0x13D, IL Offset: 0xFFFFFFFF, Source Types: StackEmpty
	Native Offset: 0x146, IL Offset: 0x74, Source Types: StackEmpty
	Native Offset: 0x149, IL Offset: 0xFFFFFFFD, Source Types: StackEmpty
	Native Offset: 0x150, IL Offset: 0xFFFFFFFE, Source Types: StackEmpty
	Native Offset: 0x166, IL Offset: 0x69, Source Types: StackEmpty
	Native Offset: 0x16D, IL Offset: 0x6C, Source Types: StackEmpty
	Native Offset: 0x17A, IL Offset: 0x6D, Source Types: CallInstruction
	Native Offset: 0x180, IL Offset: 0x72, Source Types: StackEmpty
	Native Offset: 0x181, IL Offset: 0x73, Source Types: StackEmpty
	Native Offset: 0x182, IL Offset: 0xFFFFFFFD, Source Types: StackEmpty

	Variable Locations:
	Variable Number: 1
	Start Offset: 0x22
	End Offset: 0x149
	Loc Type: VLT_STK
	Base Register: RBP
	Stack Offset: -24

	Variable Number: 2
	Start Offset: 0x22
	End Offset: 0x149
	Loc Type: VLT_STK
	Base Register: RBP
	Stack Offset: -32

	Variable Number: 3
	Start Offset: 0x22
	End Offset: 0x149
	Loc Type: VLT_STK
	Base Register: RBP
	Stack Offset: -36

	Variable Number: 0
	Start Offset: 0x22
	End Offset: 0x149
	Loc Type: VLT_STK
	Base Register: RBP
	Stack Offset: -16


    7828: 55                    push rbp
                                UWOP_PUSH_NONVOL RBP(5)
    7829: 57                    push rdi
                                UWOP_PUSH_NONVOL RDI(7)
    782a: 48 81 ec 98 00 00 00  sub rsp, 152
    7831: 48 8d ac 24 a0 00 00 00  lea rbp, [rsp + 160]
    7839: 48 8d 7d 88           lea rdi, [rbp - 120]
    783d: b9 1c 00 00 00        mov ecx, 28
    7842: 33 c0                 xor eax, eax
    7844: f3                    rep
    7845: ab                    stosd dword ptr es:[rdi], eax
    7846: 48 89 65 80           mov qword ptr [rbp - 128], rsp
    784a: 90                    nop
    784b: 48 8b 05 96 49 00 00  mov rax, qword ptr [0xc1e8]   // Dumping file: {0} (STRING_HANDLE)
    7852: 48 8b 00              mov rax, qword ptr [rax]
                                RAX is live
    7855: 48 89 45 88           mov qword ptr [rbp - 120], rax
    7859: ff 15 71 4b 00 00     call qword ptr [0xc3d0]       // Program (THREAD_STATIC_BASE_GC)
                                RAX is live
                                RAX is dead
    785f: 48 8b 10              mov rdx, qword ptr [rax]
                                RDX is live
    7862: 48 89 55 90           mov qword ptr [rbp - 112], rdx
    7866: 48 8b 55 90           mov rdx, qword ptr [rbp - 112]
    786a: 48 8b 4d 88           mov rcx, qword ptr [rbp - 120]
                                RCX is live
    786e: ff 15 24 4e 00 00     call qword ptr [0xc698]       // String System.String.Format(String, Object) (METHOD_ENTRY)
                                RAX is dead
                                RAX is live
                                RCX is dead
                                RDX is dead
    7874: 48 89 45 d0           mov qword ptr [rbp - 48], rax
    7878: 48 8b 4d d0           mov rcx, qword ptr [rbp - 48]
                                RCX is live
    787c: ff 15 1e 4e 00 00     call qword ptr [0xc6a0]       // Void System.Console.WriteLine(String) (METHOD_ENTRY)
                                RAX is dead
                                RCX is dead
    7882: 90                    nop
    7883: ff 15 47 4b 00 00     call qword ptr [0xc3d0]       // Program (THREAD_STATIC_BASE_GC)
                                RAX is live
    7889: 48 8b 00              mov rax, qword ptr [rax]
                                RAX is live
                                RAX is dead
    788c: 48 89 45 c8           mov qword ptr [rbp - 56], rax
    7890: ff 15 1a 4f 00 00     call qword ptr [0xc7b0]       // [System.Runtime]System.Text.Encoding System.Text.Encoding.get_UTF8() (METHOD_ENTRY)
    7896: 48 89 45 c0           mov qword ptr [rbp - 64], rax
    789a: ff 15 d0 4b 00 00     call qword ptr [0xc470]       // System.IO.StreamReader (NEW_OBJECT)
    78a0: 48 89 45 b8           mov qword ptr [rbp - 72], rax
    78a4: 48 8b 4d b8           mov rcx, qword ptr [rbp - 72]
                                RCX is live
    78a8: 48 8b 55 c8           mov rdx, qword ptr [rbp - 56]
                                RDX is live
    78ac: 4c 8b 45 c0           mov r8, qword ptr [rbp - 64]
                                R8 is live
    78b0: ff 15 02 4f 00 00     call qword ptr [0xc7b8]       // Void System.IO.StreamReader..ctor(String, [System.Runtime]System.Text.Encoding) (METHOD_ENTRY)
                                RAX is dead
                                RCX is dead
                                RDX is dead
                                R8 is dead
    78b6: 48 8b 4d b8           mov rcx, qword ptr [rbp - 72]
                                RCX is live
    78ba: 48 89 4d f0           mov qword ptr [rbp - 16], rcx
                                RCX is dead
    78be: 90                    nop
    78bf: 48 8b 0d 2a 49 00 00  mov rcx, qword ptr [0xc1f0]   // StreamReader created ... (STRING_HANDLE)
    78c6: 48 8b 09              mov rcx, qword ptr [rcx]
                                RCX is live
    78c9: ff 15 d1 4d 00 00     call qword ptr [0xc6a0]       // Void System.Console.WriteLine(String) (METHOD_ENTRY)
                                RCX is dead
    78cf: 90                    nop
    78d0: 48 8b 4d f0           mov rcx, qword ptr [rbp - 16]
                                RCX is live
    78d4: 4c 8d 1d 4d 4c 00 00  lea r11, [0xc528]             // String System.IO.TextReader.ReadLine() (VIRTUAL_ENTRY)
    78db: 39 09                 cmp dword ptr [rcx], ecx
    78dd: ff 15 45 4c 00 00     call qword ptr [0xc528]       // String System.IO.TextReader.ReadLine() (VIRTUAL_ENTRY)
                                RCX is dead
                                RAX is live
    78e3: 48 89 45 b0           mov qword ptr [rbp - 80], rax
    78e7: 48 8b 4d b0           mov rcx, qword ptr [rbp - 80]
                                RCX is live
    78eb: 48 89 4d e8           mov qword ptr [rbp - 24], rcx
    78ef: 48 8b 0d 02 49 00 00  mov rcx, qword ptr [0xc1f8]   // Line 1: {0} (STRING_HANDLE)
                                RCX is dead
    78f6: 48 8b 09              mov rcx, qword ptr [rcx]
                                RCX is live
    78f9: 48 8b 55 e8           mov rdx, qword ptr [rbp - 24]
                                RDX is live
    78fd: ff 15 95 4d 00 00     call qword ptr [0xc698]       // String System.String.Format(String, Object) (METHOD_ENTRY)
                                RCX is dead
                                RDX is dead
    7903: 48 89 45 a8           mov qword ptr [rbp - 88], rax
    7907: 48 8b 4d a8           mov rcx, qword ptr [rbp - 88]
                                RCX is live
    790b: ff 15 8f 4d 00 00     call qword ptr [0xc6a0]       // Void System.Console.WriteLine(String) (METHOD_ENTRY)
                                RAX is dead
                                RCX is dead
    7911: 90                    nop
    7912: 48 8b 4d f0           mov rcx, qword ptr [rbp - 16]
                                RCX is live
    7916: 4c 8d 1d 0b 4c 00 00  lea r11, [0xc528]             // String System.IO.TextReader.ReadLine() (VIRTUAL_ENTRY)
    791d: 39 09                 cmp dword ptr [rcx], ecx
    791f: ff 15 03 4c 00 00     call qword ptr [0xc528]       // String System.IO.TextReader.ReadLine() (VIRTUAL_ENTRY)
                                RCX is dead
                                RAX is live
    7925: 48 89 45 a0           mov qword ptr [rbp - 96], rax
    7929: 48 8b 4d a0           mov rcx, qword ptr [rbp - 96]
                                RCX is live
    792d: 48 89 4d e0           mov qword ptr [rbp - 32], rcx
    7931: 48 8b 0d c8 48 00 00  mov rcx, qword ptr [0xc200]   // Line 2: {0} (STRING_HANDLE)
                                RCX is dead
    7938: 48 8b 09              mov rcx, qword ptr [rcx]
                                RCX is live
    793b: 48 8b 55 e0           mov rdx, qword ptr [rbp - 32]
                                RDX is live
    793f: ff 15 53 4d 00 00     call qword ptr [0xc698]       // String System.String.Format(String, Object) (METHOD_ENTRY)
                                RCX is dead
                                RDX is dead
    7945: 48 89 45 98           mov qword ptr [rbp - 104], rax
    7949: 48 8b 4d 98           mov rcx, qword ptr [rbp - 104]
                                RCX is live
    794d: ff 15 4d 4d 00 00     call qword ptr [0xc6a0]       // Void System.Console.WriteLine(String) (METHOD_ENTRY)
                                RCX is dead
                                RAX is dead
    7953: 90                    nop
    7954: 48 83 7d e0 00        cmp qword ptr [rbp - 32], 0
    7959: 0f 95 c1              setne cl
    795c: 0f b6 c9              movzx ecx, cl
    795f: 89 4d dc              mov dword ptr [rbp - 36], ecx
    7962: 90                    nop
    7963: eb 00                 jmp 0x7965
    7965: 48 8b cc              mov rcx, rsp
    7968: e8 0b 00 00 00        call 0x7978
    796d: 90                    nop
    796e: 8b 45 dc              mov eax, dword ptr [rbp - 36]
    7971: 48 8d 65 f8           lea rsp, [rbp - 8]
    7975: 5f                    pop rdi
    7976: 5d                    pop rbp
    7977: c3                    ret
Raw Bytes:
    00007820:                         55 57 48 81 EC 98 00 00
    00007830: 00 48 8D AC 24 A0 00 00 00 48 8D 7D 88 B9 1C 00
    00007840: 00 00 33 C0 F3 AB 48 89 65 80 90 48 8B 05 96 49
    00007850: 00 00 48 8B 00 48 89 45 88 FF 15 71 4B 00 00 48
    00007860: 8B 10 48 89 55 90 48 8B 55 90 48 8B 4D 88 FF 15
    00007870: 24 4E 00 00 48 89 45 D0 48 8B 4D D0 FF 15 1E 4E
    00007880: 00 00 90 FF 15 47 4B 00 00 48 8B 00 48 89 45 C8
    00007890: FF 15 1A 4F 00 00 48 89 45 C0 FF 15 D0 4B 00 00
    000078A0: 48 89 45 B8 48 8B 4D B8 48 8B 55 C8 4C 8B 45 C0
    000078B0: FF 15 02 4F 00 00 48 8B 4D B8 48 89 4D F0 90 48
    000078C0: 8B 0D 2A 49 00 00 48 8B 09 FF 15 D1 4D 00 00 90
    000078D0: 48 8B 4D F0 4C 8D 1D 4D 4C 00 00 39 09 FF 15 45
    000078E0: 4C 00 00 48 89 45 B0 48 8B 4D B0 48 89 4D E8 48
    000078F0: 8B 0D 02 49 00 00 48 8B 09 48 8B 55 E8 FF 15 95
    00007900: 4D 00 00 48 89 45 A8 48 8B 4D A8 FF 15 8F 4D 00
    00007910: 00 90 48 8B 4D F0 4C 8D 1D 0B 4C 00 00 39 09 FF
    00007920: 15 03 4C 00 00 48 89 45 A0 48 8B 4D A0 48 89 4D
    00007930: E0 48 8B 0D C8 48 00 00 48 8B 09 48 8B 55 E0 FF
    00007940: 15 53 4D 00 00 48 89 45 98 48 8B 4D 98 FF 15 4D
    00007950: 4D 00 00 90 48 83 7D E0 00 0F 95 C1 0F B6 C9 89
    00007960: 4D DC 90 EB 00 48 8B CC E8 0B 00 00 00 90 8B 45
    00007970: DC 48 8D 65 F8 5F 5D C3

Program.StreamReaderReadLine()
Id: 25
StartAddress: 0x00007978
Size: 57 bytes
UnwindRVA: 0x0000A65C
Version:            1
Flags:              0x03 EHANDLER UHANDLER
SizeOfProlog:       0x0006
CountOfUnwindCodes: 3
FrameRegister:      RAX
FrameOffset:        0x0
PersonalityRVA:     0x6000
UnwindCode[0]: CodeOffset 0x0006 FrameOffset 0x4206 NextOffset 0x-1 Op 40
UnwindCode[1]: CodeOffset 0x0002 FrameOffset 0x7002 NextOffset 0x-1 Op RDI(7)
UnwindCode[2]: CodeOffset 0x0001 FrameOffset 0x5001 NextOffset 0x-1 Op RBP(5)

    7978: 55                    push rbp
    7979: 57                    push rdi
    797a: 48 83 ec 28           sub rsp, 40
    797e: 48 8b 69 20           mov rbp, qword ptr [rcx + 32]
    7982: 48 89 6c 24 20        mov qword ptr [rsp + 32], rbp
    7987: 48 8d ad a0 00 00 00  lea rbp, [rbp + 160]
    798e: 48 83 7d f0 00        cmp qword ptr [rbp - 16], 0
    7993: 74 14                 je 0x79a9
    7995: 48 8b 4d f0           mov rcx, qword ptr [rbp - 16]
    7999: 4c 8d 1d 90 4b 00 00  lea r11, [0xc530]             // Void System.IDisposable.Dispose() (VIRTUAL_ENTRY)
    79a0: 39 09                 cmp dword ptr [rcx], ecx
    79a2: ff 15 88 4b 00 00     call qword ptr [0xc530]       // Void System.IDisposable.Dispose() (VIRTUAL_ENTRY)
    79a8: 90                    nop
    79a9: 90                    nop
    79aa: 48 83 c4 28           add rsp, 40
    79ae: 5f                    pop rdi
    79af: 5d                    pop rbp
    79b0: c3                    ret
Raw Bytes:
    00007970:                         55 57 48 83 EC 28 48 8B
    00007980: 69 20 48 89 6C 24 20 48 8D AD A0 00 00 00 48 83
    00007990: 7D F0 00 74 14 48 8B 4D F0 4C 8D 1D 90 4B 00 00
    000079A0: 39 09 FF 15 88 4B 00 00 90 90 48 83 C4 28 5F 5D
    000079B0: C3

_______________________________________________
</pre>

trylek commented:
Awesome, thanks so much @BruceForstall for your valuable feedback. I guess as the next step I'll need to look into the live / dead status transitions as we discussed before.
